feat: support jsonrpc 0.2.1

Matthew-Herman opened the pull request:
These changes add support for the jsonrpc 0.2.1 spec for [issue 244](https://github.com/xJonathanLEI/starknet-rs/issues/244).

xJonathanLEI commented on starknet-providers/src/jsonrpc/mod.rs:
```diff
@@ -111,8 +111,10 @@ struct FeltArray(#[serde_as(as = "Vec<UfeHex>")] pub Vec<FieldElement>);
     filter: EventFilter,
-    page_size: u64,
-    page_number: u64,
+    /// a pointer to the last element of the delivered page, use this token in a subsequent query to obtain the next page
```
No comment/docs lines should exceed 100 chars for readability. Hard-wrap it into the next line. Please also fix other occurrences.

xJonathanLEI commented on starknet-providers/src/jsonrpc/mod.rs:
```diff
@@ -224,11 +226,15 @@ where
     /// Get the contract class definition associated with the given hash
```
The description for this method has changed to `Get the contract class definition in the given block associated with the given hash` in the specs. Please reflect it here in the docs. Please also check for other desc changes.

xJonathanLEI commented on starknet-providers/src/jsonrpc/mod.rs:
```diff
@@ -301,18 +307,15 @@ where
-        request: R,
+        request: &BroadcastedTransaction,
         block_id: &BlockId,
-    ) -> Result<FeeEstimate, JsonRpcClientError<T::Error>>
-    where
-        R: AsRef<FunctionCall>,
-    {
+    ) -> Result<FeeEstimate, JsonRpcClientError<T::Error>> {
         self.send_request(
```
The use of generic here is by design. In some cases it's impossible to pass references since the value gets dropped immediately. Using `AsRef` generic here allows the function to be used in those cases.

Extra caution is needed if you're changing the design.

Matthew-Herman replied:
Reverted to using the generic for estimate_fee

xJonathanLEI commented on starknet-providers/src/jsonrpc/models/mod.rs:
```diff
@@ -591,6 +594,228 @@ pub struct BlockHashAndNumber {
+
+/// (`CONTRACT_ABI`)
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct ContractABI {
```
Rename all `ABI` to `Abi`

xJonathanLEI commented on starknet-providers/src/jsonrpc/mod.rs:
```diff
@@ -388,53 +400,35 @@ where
+            [serde_json::to_value(deploy_transaction)?],
         )
         .await
     }
```
Missing the new `starknet_addDeployAccountTransaction` method?

xJonathanLEI commented on starknet-providers/src/jsonrpc/models/mod.rs:
```diff
@@ -286,14 +284,15 @@ pub struct StorageDiffItem {
+/// Transaction (`TXN`) The transaction schema, as it appears inside a block
 #[derive(Debug, Clone, Serialize, Deserialize)]
-#[serde(tag = "type", rename_all = "SCREAMING_SNAKE_CASE")]
+#[serde(untagged)]
```
The change to `untagged` here seems unjustified. It shouldn't be necessary for making things work, since the raw data transferred over the wire _does_ come tagged.

xJonathanLEI commented on starknet-providers/src/jsonrpc/models/mod.rs:
```diff
@@ -396,47 +388,64 @@ pub struct TransactionReceiptMeta {
-/// The `TXN_RECEIPT` type in the specification, except without the `PENDING_TXN_RECEIPT` variant.
+/// (`TXN_RECEIPT`)
 #[derive(Debug, Clone, Serialize, Deserialize)]
 #[serde(untagged)]
```
This `untagged` used to be necessary, but it looks like the new spec version does tag the receipts, making the use of `untagged` here unnecessary. Change it to tagged.

xJonathanLEI replied:
`untagged`s on all receipt types should be reviewed as they should no longer be necessary.

xJonathanLEI commented on starknet-providers/src/jsonrpc/models/mod.rs:
```diff
@@ -591,6 +594,228 @@ pub struct BlockHashAndNumber {
+
+/// (`CONTRACT_ABI_ENTRY`)
+#[derive(Debug, Clone, Serialize, Deserialize)]
+#[serde(untagged)]
```
This should not be untagged. It's tagged in the spec.

xJonathanLEI reviewed (changes requested):
Apart from the comments already made, looks like `untagged` is misused across the PR. Please review the use of `untagged` and change accordingly.

Matthew-Herman commented:
Tried to remove use of untagged where possible. Not sure if there is a way to avoid using `.clone()` in the add transaction methods? I added it because initialising the enum there to add the tag takes ownership of the method param.

fracek commented:
FYI, this patch is needed to use the json rpc provider with infura. Any way I can help to get this merged?

xJonathanLEI commented:
@fracek ima get it done tmr

Matthew-Herman commented:
Removed the tests for add_deployTransaction because StarkNet v0.10.3 removes the deploy transaction and was deployed to goerli yesterday. Also removed tests for adding version 0 transactions because they are deprecated on the docs.

xJonathanLEI commented on starknet-providers/src/jsonrpc/models/mod.rs:
```diff
@@ -18,21 +18,15 @@ pub enum MaybePendingBlockWithTxHashes {
 }
 
-#[derive(Debug, Clone, Serialize, Deserialize)]
+#[derive(Debug, Clone, Deserialize)]
```
There are many occurrences where the `Serialize` derive gets removed by this PR. Please don't do this. Technically we don't need serialization for the JSON-RPC client to work. However, as an SDK we need to consider use cases of dependent libs/apps, who are very likely to need serialization to function properly.

xJonathanLEI replied:
Serialize trait derives added back.

xJonathanLEI commented on starknet-providers/src/jsonrpc/mod.rs:
```diff
@@ -388,53 +393,56 @@ where
+        invoke_transaction: &BroadcastedInvokeTransaction,
     ) -> Result<InvokeTransactionResult, JsonRpcClientError<T::Error>> {
         self.send_request(
             JsonRpcMethod::AddInvokeTransaction,
-            [
-                serde_json::to_value(function_invocation)?,
-                serde_json::to_value(FeltArray(signature))?,
-                serde_json::to_value(Felt(max_fee))?,
-                serde_json::to_value(Felt(version))?,
-            ],
+            [serde_json::to_value(BroadcastedTransaction::Invoke(
+                invoke_transaction.clone(),
+            ))?],
         )
         .await
     }
```
Asking the user to pass in reference only to `clone()` inside is bad practice. The user might have an object he's willing to give up ownership for, but this function here is forcing an extra clone, which could be expensive.

The responsibility for cloning is on the user, _if_ we have to take ownership.

xJonathanLEI replied:
Actually, there's no need to take ownership here at all. You shouldn't need to wrap the object in a `BroadcastedTransaction` enum in the first place. It's just for serialization after all.

Matthew-Herman replied:
Moved tagging with type from the enum to structs as it is only used for serialization

xJonathanLEI commented on starknet-providers/src/jsonrpc/models/mod.rs:
```diff
@@ -591,6 +681,196 @@ pub struct BlockHashAndNumber {
+    pub constructor_calldata: Vec<FieldElement>,
+}
+
+/// (`BROADCASTED_DEPLOY_ACCOUNT_TXN`) Mempool representation of a deploy account transaction
+#[serde_as]
+#[derive(Debug, Clone, Serialize)]
+pub struct BroadcastedDeployAccountTransaction {
+    /// The maximal fee that can be charged for including the transaction
+    #[serde_as(as = "UfeHex")]
+    pub max_fee: FieldElement,
+    /// Version of the transaction scheme
+    pub version: TransactionVersion,
+    #[serde_as(as = "Vec<UfeHex>")]
+    pub signature: Vec<FieldElement>,
+    #[serde_as(as = "UfeHex")]
+    pub nonce: FieldElement,
+    /// The salt for the address of the deployed contract
+    #[serde_as(as = "UfeHex")]
+    pub contract_address_salt: FieldElement,
```
I don't like how this allows people to construct invalid transaction requests: you can specify `version` to be `V0` without encountering a compile-time error. However there's actually no v0 for `DeployAccount` transaction type.

Matthew-Herman replied:
Understand we should do this. I'm a bit confused about which transaction versions the types should allow to be constructed here. Not sure the [docs.starknet.io](https://docs.starknet.io/documentation/architecture_and_concepts/Blocks/transactions/#declare_transaction) and jsonrpc spec are giving me a full picture. Should this just target v0.10.3 and should constructing transaction version 0 for declare/invoke/deploy transactions be allowed? And also now that adding deploy transactions are deprecated from v0.10.3, should the add_deploy_transaction method be removed?

xJonathanLEI replied:
I general we don't support deprecated features since they will be removed from the protocol completely (without needing to support old block data) eventually with the regenesis. However, I want to make an exception here for JSON RPC as we're specifically targeting a certain specification. Making our code closer to the specs helps with maintenance.

xJonathanLEI commented:
Superseded by #255.
